sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GlobalAddressResolver.java

```diff
@@ -94,69 +97,142 @@ public GlobalAddressResolver(
+                                                        hint
+                                                );
+                                            });
+                                }), Configs.getCPUCnt(), Configs.getCPUCnt());
     }
 
-    private Flux<OpenConnectionResponse> openConnectionsAndInitCachesInternal(
+    private Flux<ImmutablePair<ImmutablePair<String, DocumentCollection>, AddressInformation>> resolveAddressesPerCollection(
+            String containerLink,
             DocumentCollection collection,
             List<PartitionKeyRangeIdentity> partitionKeyRangeIdentities,
-            CosmosContainerProactiveInitConfig proactiveContainerInitConfig
+            CosmosContainerProactiveInitConfig proactiveContainerInitConfig,
+            AsyncDocumentClient.OpenConnectionAggressivenessHint hint
     ) {
+        if (proactiveContainerInitConfig.getProactiveConnectionRegionsCount() > 0) {
+            return Flux.fromStream(this.endpointManager.getReadEndpoints().stream())
+                    .take(proactiveContainerInitConfig.getProactiveConnectionRegionsCount())
+                    .flatMap(readEndpoint -> {
+                        if (this.addressCacheByEndpoint.containsKey(readEndpoint)) {
+                            return this.addressCacheByEndpoint.get(readEndpoint)
+                                    .addressCache
+                                    .resolveAddressesAndInitCaches(
+                                            containerLink,
+                                            collection,
+                                            partitionKeyRangeIdentities,
+                                            hint
+                                    );
+                        }
+                        return Flux.empty();
+                    });
+        }
+
+        return Flux.empty();
+    }
 
+    private Flux<Void> submitOpenConnectionInternal(
+            AddressInformation address,
+            DocumentCollection documentCollection,
+            CosmosContainerProactiveInitConfig proactiveContainerInitConfig,
+            int connectionPerEndpointCount,
+            AsyncDocumentClient.OpenConnectionAggressivenessHint hint
+    ) {
         if (proactiveContainerInitConfig.getProactiveConnectionRegionsCount() > 0) {
             return Flux.fromStream(this.endpointManager.getReadEndpoints().stream())
                     .take(proactiveContainerInitConfig.getProactiveConnectionRegionsCount())
                     .flatMap(readEndpoint -> {
                         if (this.addressCacheByEndpoint.containsKey(readEndpoint)) {
                             return this.addressCacheByEndpoint.get(readEndpoint)
                                     .addressCache
-                                    .openConnectionsAndInitCaches(collection, partitionKeyRangeIdentities);
+                                    .submitOpenConnectionTask(
```

Comment:
how do you know the address belongs to this region?

I mean during `resolveAddressesPerCollection` you resolved all the addresses from all regions
region A -> address1, address2
region B -> address3, address4
so here you are going to submit 8 openConnection task instead of 4?
regionA -> address1, address2, address3, address4
regionB -> address1, address2, address3, address4